avm/src/daemon/daemon.rs

```diff
@@ -173,43 +206,112 @@ pub async fn start(
   cert_b64: Option<&str>,
 ) {
   let cluster_id = cluster_id.to_string();
+  CLUSTER_ID.set(String::from(&cluster_id)).unwrap();
   let deploy_token = deploy_token.to_string();
   let agzb64 = agz_b64.to_string();
   let domain = domain.to_string();
   task::spawn(async move {
     // TODO even better period determination
     let period = Duration::from_secs(5 * 60);
-    let dns = DNS::new(&domain);
     let self_ip = get_private_ip().await;
     let mut cluster_size = 0;
     let mut leader_ip = "".to_string();
-    loop {
-      sleep(period).await;
-      let vms = dns.get_vms(&cluster_id).await;
-      // triggered the first time since cluster_size == 0
-      // and every time the cluster changes size
-      if vms.len() != cluster_size {
-        cluster_size = vms.len();
-        let ips = vms
-          .iter()
-          .map(|vm| vm.private_ip_addr.to_string())
-          .collect();
-        let lrh = LogRendezvousHash::new(ips);
-        leader_ip = lrh.get_leader_id().to_string();
+    let dns = DNS::new(&domain);
+    match (dns, self_ip) {
+      (Ok(dns), Ok(self_ip)) => {
+        loop {
+          sleep(period).await;
+          let mut vms_err: Option<String> = None;
+          let vms = dns.get_vms(&cluster_id).await.unwrap_or_else(|err| {
+            vms_err = Some(format!("{}", err));
+            return Vec::new();
+          });
+          if let Some(err) = vms_err {
+            error!(114, "{}", err).await;
+          };
+          // triggered the first time since cluster_size == 0
+          // and every time the cluster changes size
+          if vms.len() != cluster_size {
+            cluster_size = vms.len();
+            let ips = vms
+              .iter()
+              .map(|vm| vm.private_ip_addr.to_string())
+              .collect();
+            let lrh = LogRendezvousHash::new(ips);
+            leader_ip = lrh.get_leader_id().to_string();
+          }
+          if leader_ip == self_ip {
+            let mut factor_err: Option<String> = None;
+            let factor = post_v1_stats(&cluster_id, &deploy_token)
+              .await
+              .unwrap_or_else(|err| {
+                factor_err = Some(format!("{}", err));
+                return "1".to_string();
+              });
+            if let Some(err) = factor_err {
+              error!(115, "{}", err).await;
+            };
+            println!(
+              "VM stats sent for cluster {} of size {}. Cluster factor: {}.",
+              cluster_id,
+              vms.len(),
+              factor
+            );
+            if factor != "1" {
+              post_v1_scale(&cluster_id, &agzb64, &deploy_token, &factor).await;
+            }
+          }
+        }
+      }
+      (Err(dns_err), Ok(_self_ip)) => {
```

Comment:
These changes are for the best, but they are going to make it harder to locally test `daemon` mode changes. Need to think about how to handle that.